nexus/db-queries/src/db/pool_connection.rs

```diff
@@ -224,90 +39,152 @@ impl CustomizeConnection<Connection<DbConnection>, ConnectionError>
+    use crate::db::datastore::test_utils::datastore_test;
+    use async_bb8_diesel::AsyncRunQueryDsl;
+    use nexus_db_model::AllSchemaVersions;
+    use nexus_db_model::Resources;
+    use nexus_db_model::SledResource;
+    use nexus_db_model::SledResourceKind;
+    use nexus_db_model::EARLIEST_SUPPORTED_VERSION;
+    use nexus_db_model::SCHEMA_DIR;
+    use nexus_test_utils::db::test_setup_database_empty;
     use omicron_test_utils::dev;
+    use slog::info;
+    use std::sync::Arc;
+    use std::time::Duration;
+    use uuid::Uuid;
 
-    // Ensure that the "CUSTOM_TYPE_KEYS" values match the enums
-    // we find within the database.
-    //
-    // If the two are out-of-sync, identify the values causing problems.
     #[tokio::test]
-    async fn all_enums_in_prepopulate_list() {
-        let logctx = dev::test_setup_log("test_project_creation");
-        let mut crdb = test_setup_database(&logctx.log).await;
-        let client = crdb.connect().await.expect("Failed to connect to CRDB");
+    async fn diesel_oid_cache_isnt_poisoned_after_schema_upgrade() {
+        let logctx = dev::test_setup_log(
+            "diesel_oid_cache_isnt_poisoned_after_schema_upgrade",
+        );
 
-        // https://www.cockroachlabs.com/docs/stable/show-enums
-        let rows = client
-            .query("SHOW ENUMS FROM omicron.public;", &[])
-            .await
-            .unwrap_or_else(|_| panic!("failed to list enums"));
-        client.cleanup().await.expect("cleaning up after listing enums");
+        // Start with an empty DB.
+        let db = Arc::new(test_setup_database_empty(&logctx.log).await);
+
+        // On a background task, start creating a datastore. This will block
+        // until the schema version matches what it expects (the latest
+        // version), but we want to go ahead and create it so it creates a
+        // connection pool and starts polling CRDB. If at any point in this
+        // process we populate diesel's OID cache, it will be poisoned by any
+        // OID changes caused by our schema migrations.
+        let datastore_task = tokio::spawn({
+            let logctx = dev::test_setup_log(
+                "diesel_oid_cache_isnt_poisoned_after_schema_upgrade_datastore",
+            );
+            let db = Arc::clone(&db);
+            async move {
+                let (_opctx, datastore) = datastore_test(&logctx, &db).await;
+                datastore
+            }
+        });
+
+        // Apply all the schema migrations in order.
+        {
+            let client = db.connect().await.expect("Failed to connect to CRDB");
+            client
+                .batch_execute(DISALLOW_FULL_TABLE_SCAN_SQL)
+                .await
+                .expect("failed to disallow full table scans");
+
+            let all_versions = AllSchemaVersions::load(SCHEMA_DIR.into())
+                .expect("Failed to load schema versions");
+            for version in all_versions.iter_versions() {
```

Comment:
It does make me a little nervous to duplicate this logic.  I wonder if we should modify the existing tests that do this instead?

Reply:
(This one's not a blocker for me.)

Reply:
The existing tests that do this don't use a connection pool. I guess they could? I'll look and see (taking the other comment into consideration too while I'm there).